### PR TITLE
Simplification of GenericPowerModel{T} Type

### PR DESCRIPTION
**Pull request by ccoffrin**
This is a fairly speculative PR aimed at addressing issue #192.  The basic idea is replace `GenericPowerModel{T}` with named concrete types `NAMEPowerModel` which are all a subtype of `AbstractPowerModel`.  The key insight is to use the macro, `@pm_fields` to provide a uniform set of typed fields in the concrete `NAMEPowerModel` types.  For example, this is how typical types are defined,
```
abstract type AbstractPowerModel end

abstract type AbstractACPModel <: AbstractPowerModel end
mutable struct ACPPowerModel <: AbstractACPModel @pm_fields end

abstract type AbstractACRModel <: AbstractPowerModel end
mutable struct ACRPowerModel <: AbstractACRModel @pm_fields end
```
Because the types have a consistent set of fields the share function `InitializePowerModel` can be used to initialize any struct specified with `@pm_fields`.

The removal of the type parameter makes function overloading cleaner,
```
before: function some_fun(pm::GenericPowerModel{T}, ...) where T <: AbstractACPModel end

after: function some_fun(pm::AbstractACPModel, ...) end
```

Overall this approach seems to have the following advantages,
- easier to read, understand and extend the code
- No need to "Form" and "Model" objects that are seemingly the same.
- no need to overload type constructors when defining new types, `InitializePowerModel` will be used

The primary disadvantage is that it requires rewriting the signatures of nearly all of the downstream functions.  Easy but tedious work.

If we generally like this new design, then similar updates should made across all of InfrastructureModels.  Thoughts @rb004f @kaarthiksundar @pseudocubic @tasseff?

Open questions:
- Does this structure create any challenges in building package extensions?
- Will this make the multi-infrastructure packages easier or harder to build?  Is there a simple feature to add that makes supporting multi-infrastructure easy?

CC @jd-lara, @frederikgeth, @sanderclaeys

### CODE REVIEW

**Comment by sanderclaeys** [comment]
Apart from multi-infrastructure, it would be great to see explicit support for 'multi-component'. 

Take for example OLTCs and delta connections in TPPM. Each of these would have at least two formulations in the W variable space. The current formulation actually refers to the formulation of the branches mostly. Creating a new named formulation explicitly for every possible combination, {oltc1, oltc2}x{delta1, delta2}x{branch1, branch2, branch3...} is not so scalable in the long run. A workaround we use now is creating new problem definitions for what are actually variations of component formulations.

One design I could think of for this, is allowing parallel type hierarchies for these new components. To hide this complexity from the new user, we could define default formulations for the components, depending on the choice of branch formulation. For example, an OLTC might have two different convex relaxations in the W variable space. When the user chooses a WR (branch) formulation, a default choice for the OLTC formulation is made, compatible with this branch formulation.

This might not qualify as a simple feature though.

**Comment by ccoffrin** [comment]
@sanderclaeys, you make a good point, this is an issue that we need to sort out eventually.  However, I do not immediately see how the issue you point out is directly relevant to this PR, did I miss something?

Some guidelines for AbstractPowerModel types that may help in the near term,
- As of now, the one-and-only role of the AbstractPowerModel types are to select different possible implementations of a set of fixed component models and problem constraints+objective.
- The AbstractPowerModel types are basically 1-to-1 with variable spaces.  If two formulations have different variable spaces they need different types, even if they can be proved to be equivalent (e.g. BIM and BFM being the canonical example).
- If you are changing component modeling assumptions, this should be handled through different constraint names and consequently different problem formulations (it seems this is what you are doing already). Generalizing over these component and problem definitions is the goal of tools on top of PowerModels, e.g. PowerSimulations.

**Comment by sanderclaeys** [comment]
It is indeed beyond the scope of this PR; I do not think it should be addressed right now. I thought this was a good moment to raise that concern however, as it would also involve reorganizing how the functions are overloaded on formulation type.

For the OLTC for example, there are some papers that only link the diagonal of W, whilst others also link the off-diagonals. This is not a generalization but different ways to relax a basic, low-level component. In that sense it would be nice to have a design where we can overload the functions for the OLTC in a similar way as is done for the branches, but specifically on an 'OLTC' formulation type, which can refer back to the paper where it is introduced.

**Comment by frederikgeth** [comment]
This does look cleaner and I think it makes it easier for new contributors to get started. Don't really have an opinion on the open questions.

I subscribe to @sanderclaeys  point. Power system modellers develop hierarchies of formulations for transformers, on-load tap changers, voltage regulators, dc branches, HVDC converter stations, battery storage, etc. At the moment this is a bit inflexible. Maybe we should plan some discussions with the people that attend PES GM?

**Comment by ccoffrin** [comment]
@frederikgeth, agreed the GM would be a place to have that discussion and develop a good software design for this.

**Comment by tasseff** [comment]
Am I correct that this will reduce the ability to define functions that are valid for a union of types? The alternative suggests that a more fixed model hierarchy will need to be used. As an example, without these changes, this can be accomplished with
```julia
function fun(pm::GenericPowerModel{T}, ...) where T <: Union{AbstractModel1, AbstractModel2} 
```
but will now require `AbstractModel1` and `AbstractModel2` to be inherited from some type `AbstractModelMaster`, where the above would, with these changes, now be rewritten as
```julia
function fun(pm::AbstractModelMaster, ...)
```
Am I missing something? If not, I'm actually in favor of keeping the type parameter in function headers. I think the benefits of flexible typing, without the overhead of needing to create model hierarchies, outweigh the benefits of smaller function headers. I am open to persuasion, though.

**Comment by pseudocubic** [comment]
@tasseff I believe you would just define your function like this:

`function fun(pm::Union{AbstractModel1, AbstractModel2}, ...)`

File: src/core/data.jl [hunk 1]
```diff
@@ -231,7 +231,7 @@ function apply_func(data::Dict{String,<:Any}, key::String, func)
 end
 
 
-"Transforms network data into per-unit"
+"TransModels network data into per-unit"
```
Comment (pseudocubic) [inline src/core/data.jl]: Overzealous search and replace

**Comment by pseudocubic** [comment]
@ccoffrin I think this would be a good change. From the perspective of the packages I maintain, the downstream changes would be tedious, as you said, but not especially difficult, and I think this greatly improves readability and makes adding new formulations much cleaner, so any pain in upgrading would be worth it to me.

**Comment by jd-lara** [comment]
The main changes in the PR won't affect too much how we interface with PowerModels in PowerSimulations. It might require some code rewrite but nothing significant. 

To @tasseff point, I think that future implementations of the approach in this PR might require new abstractions to avoid the use of Unions and enable more code re-use. 

As far as I understand the "Julia's ways" of doing things, most functions will call on an abstract type and then users can define their new types under the abstraction tree and be able to reuse a lot of the code and this implementation moves towards that kind of approach. 

The issues that @sanderclaeys mentions are what we are trying to tackle in PowerSimulations where users need a heterogeneous collection of devices in the model exactly as @frederikgeth mentioned it. Implementing this and interacting with PowerModels is something I am working on right now, finding the right balance is non-trivial but unrelated to this improvement.

**Comment by ccoffrin** [comment]
Overall it seems like folks are generally in agreement with this notable update and the required maintenance burden.  I will test that the PowerModelsAnnex update is reasonable before finalizing the change; this change is planned for v0.13, after some other breaking changes.

@tasseff, David's suggestion is correct.  This should not change our ability to use type unions when defining what functions apply to what models.

@jd-lara, BTW, I looked into this update at your request.  I think it was becouse the Model vs Form duplication was causing confusion of users at the PowerSimulations level?  If we make this change will PowerSimulation's needs be met?

**Comment by jd-lara** [comment]
@ccoffrin yes, this will eliminate the confusion in our users between Models and Form. Thanks for the update. 

On a side note, it will be good to add to the feature request (I can make a PR later with a proposal too) of initializing a PowerModels and later on adding data to it so we don't have to rely on Anonymous functions for the interface.

**Comment by ccoffrin** [comment]
Great!  Can you open an issue with the feature request for initializing a PowerModel without data?  I am not following your brief description, and example in a issue would be great.

**Comment by tasseff** [comment]
@ccoffrin, @pseudocubic Thanks for elaborating on this. In that case, I definitely agree with these changes.

**Comment by ccoffrin** [comment]
Note to self, make sure to explore the generic API idea discussed in #192.

**Pull request merged**